mds/src/karatsuba_convolution.rs

```diff
@@ -0,0 +1,362 @@
+/// improvements.
+///
+/// NB: If primitive multiplications are still the bottleneck, a
+/// futher possibility would be to find an MDS matrix some of whose
+/// entries are powers of 2. Then the multiplication can be replaced
+/// with a shift, which on most architectures has better throughput
+/// and latency, and is issued on different ports (1*p06) to
+/// multiplication (1*p1).
+pub trait Convolve<F, T: RngElt, U: RngElt, V: RngElt> {
+    /// Given an input element, retrieve the corresponding internal
+    /// element that will be used in calculations.
+    fn read(input: F) -> T;
+
+    /// Given input vectors `lhs` and `rhs`, calculate their dot
+    /// product. The result can be reduced with respect to the modulus
+    /// (of `F`), but it must have the same lower 10 bits as the dot
+    /// product if all inputs are considered integers. See
+    /// `baby-bear/src/mds.rs::barret_red_babybear()` for an example
+    /// of how this can be implemented in practice.
+    fn parity_dot<const N: usize>(lhs: [T; N], rhs: [U; N]) -> V;
+
+    /// Convert an internal element of type `V` back into an external
+    /// element.
+    fn reduce(z: V) -> F;
+
+    /// Convolve `lhs` and `rhs`.
+    ///
+    /// The parameter `conv` should be the function in this trait that
+    /// corresponds to length `N`.
+    #[inline(always)]
+    fn apply<const N: usize, C: Fn([T; N], [U; N], &mut [V])>(
+        lhs: [F; N],
+        rhs: [U; N],
+        conv: C,
+    ) -> [F; N] {
+        let lhs = lhs.map(Self::read);
+        let mut output = [V::default(); N];
+        conv(lhs, rhs, &mut output);
+        output.map(Self::reduce)
+    }
+
+    /// Compute output(x) = lhs(x)rhs(x) mod x^N - 1.
+    ///
+    /// NB: This function should not be called by clients; use `Self::apply()`.
+    #[inline(always)]
+    fn conv_n<
+        const N: usize,
+        const HALF_N: usize,
+        C: Fn([T; HALF_N], [U; HALF_N], &mut [V]),
+        SC: Fn([T; HALF_N], [U; HALF_N], &mut [V]),
+    >(
+        lhs: [T; N],
+        rhs: [U; N],
+        output: &mut [V],
+        inner_conv: C,
+        inner_signed_conv: SC,
+    ) {
+        debug_assert_eq!(2 * HALF_N, N);
+        // NB: The compiler is smart enough not to initialise these arrays.
+        let mut lhs_pos = [T::default(); HALF_N]; // lhs_pos = lhs(x) mod x^{N/2} - 1
+        let mut lhs_neg = [T::default(); HALF_N]; // lhs_neg = lhs(x) mod x^{N/2} + 1
+        let mut rhs_pos = [U::default(); HALF_N]; // rhs_pos = lhs(x) mod x^{N/2} - 1
+        let mut rhs_neg = [U::default(); HALF_N]; // rhs_neg = lhs(x) mod x^{N/2} + 1
+
+        for i in 0..HALF_N {
+            let s = lhs[i];
+            let t = lhs[i + HALF_N];
+
+            lhs_pos[i] = s + t;
+            lhs_neg[i] = s - t;
+
+            let s = rhs[i];
+            let t = rhs[i + HALF_N];
+
+            rhs_pos[i] = s + t;
+            rhs_neg[i] = s - t;
+        }
+
+        let (left, right) = output.split_at_mut(HALF_N);
+
```

Comment:
To make the comment in lines 192 - 194 make more sense, might be worth changing the 185, 188 comments to

left = w1 = lhs(x)rhs(x) mod x^{N/2} + 1
right = w0 = lhs(x)rhs(x) mod x^{N/2} - 1

Or something to that effect.